src/adagrams.js

```diff
@@ -1,15 +1,184 @@
+    const element = letterBox[random];
+    letterList.push(element);
+    letterBox.splice(random, 1);
+    
+  }
+  return letterList;
+  
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+
+  const letterFrequency = {};
+
+  for (let index = 0; index < lettersInHand.length; index++) {
+    const element = lettersInHand[index].toLowerCase();
+    //gets the current value, if doesnt exist, return 0, if it exists add 1 to the current value
+    letterFrequency[element] = (letterFrequency[element] || 0) + 1;
+  }
+
+
+  for (let index = 0; index < input.length; index++) {
+    const element = input[index].toLowerCase();
+    if (!letterFrequency[element]) {
+      return false;
+      
+    }
+    //keep track of available letters
+    else {
+      letterFrequency[element] -= 1;
+    }
+    
+  }
+  
+  return true;
+  
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  
+  let wordScore = 0
+
+  const letterSet1 = ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"];
+  const letterSet2 = ["D", "G"];
+  const letterSet3 = ["B", "C", "M", "P"];
+  const letterSet4 = ["F", "H", "V", "W", "Y"];
+  const letterSet5 = ["K"];
+  const letterSet6 = ["J", "X"];
+  const letterSet7 = ["Q", "Z"];
+
+  const casedWord = word.toUpperCase();
+  //loop through the word and check if any of the letters are in sets and update the score
+
+  if (word === "") {
+    return 0;
+  }
+
+  for (let letter of casedWord) {
+    if (letterSet1.includes(letter)) {
+      wordScore += 1;
+    }
+    if (letterSet2.includes(letter)) {
+      wordScore += 2;
+    }
+    if (letterSet3.includes(letter)) {
+      wordScore += 3;
+    }
+    if (letterSet4.includes(letter)) {
+      wordScore += 4;
+    }
+    if (letterSet5.includes(letter)) {
+      wordScore += 5;
+    }
+    if (letterSet6.includes(letter)) {
+      wordScore += 8;
+    }
+    if (letterSet7.includes(letter)) {
+      wordScore += 10;
+    }
+  }
+
+  if (casedWord.length >= 7) {
+    wordScore += 8;
+  }
+  return wordScore;
+
 };
 
+  
+
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
+
+  //pseudocode requirement wave 4
+  
+  //initialize variables (maxScore, bestWord, currentScore)
+  //loop through the words and use the helper function - scoreWord to extract the score and pass it to currentScore
+    // checkpoint - if currentScore > maxScore, currentScore becomes the new maxScore and current word in the loop becomes the bestWord
+    // if the currentScore === maxScore:
+          // check if the length of the word === 10 and the length of best word != 10, word becomes the bestWord
+          // if the word < the length of bestWord and the length of bestWord < 10, word becomes the new bestWord
+  
+    //finally return bestWord, and maxScore
+  
+  
+  let maxScore = 0;
+  let bestWord = "";
+  let currentScore = 0;
```

Comment:
Small nitpick here, but `currentScore` does not need to be declared here. It can be declared inside the for loop. Since it's being reassigned at each iteration, it can also be declared at each iteration since we do not need to save the value from the previous iteration.